Use new API for label scans

pontusmelke opened the pull request:
This PR accomplishes the following things:

1. Makes label scans transaction-state-aware in new API
2. Use the new API in interpreted runtime(s)
3. Use the new API in compiled runtime
4. Use the new API in Core API

pontusmelke committed:
Remove unnecessary set

There is no reason to store an extra copy of the removed nodes, instead
delegate to the tx state to do the check.

pontusmelke committed:
Extra method instead of base implementation of next

pontusmelke committed:
Support for multiple labels in tx state

Added the ability to query the transaction state for changes in multiple
labels.

pontusmelke committed:
Use API for label scan in interpreted runtime

pontusmelke committed:
Use API for label scan in compiled runtime

pontusmelke committed:
Use API for label scan in core api

fickludd commented on community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/CantCompileQueryException.scala:
```diff
@@ -25,5 +25,6 @@ import org.neo4j.cypher.internal.util.v3_4.spi.MapToPublicExceptions
 class CantCompileQueryException(message: String = "Internal error - should have used fall back to execute query, but something went horribly wrong", cause:Throwable=null)
   extends CypherException(message, cause) {
 
+  if (cause != null) cause.printStackTrace()
```
Whoops

fickludd commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/TransactionStateTestBase.java:
```diff
@@ -518,6 +514,194 @@ public void shouldNotSeeDeletedNode() throws Exception
+              NodeLabelIndexCursor cursor = cursors.allocateNodeLabelIndexCursor() )
+        {
+            // when
+            tx.dataWrite().nodeAddLabel( node.node, node.labels[0] );
```
Did you mean to re-assign the same label to the node, or should it be created with no label?

fickludd commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/TransactionStateTestBase.java:
```diff
@@ -518,6 +514,194 @@ public void shouldNotSeeDeletedNode() throws Exception
+        {
+            // when
+            tx.dataWrite().nodeRemoveLabel( node.node, node.labels[1] );
+            tx.dataRead().nodeLabelUnionScan( cursor, node.labels );
```
Maybe I'm confused. I though the Union scan should return all nodes with a least one of the provided labels? In that we should still find node as it retains `:label1`.

pontusmelke replied:
No the confusion is on me. I disabled `nodeIntersectionScan` for now since it is tricky to combine with tx state and currently not used.

fickludd reviewed (changes requested):
I'm slightly unsure about the `nodeLabelUnionScan` so would like to discuss that before merging. It seems to behave the same as `nodeLabelIntersectScan` to me, but maybe I'm misreading somehow. I guess we don't really have any way to benefit from scanning over multiple layers yet anyway, as the label scan store is on single label? So we could just slam `UnsupportedOperationException` on these methods until the kernel can provide a more effective implementation or someone want to use them in the runtime.

Otherwise great test coverage and nice implementation.

pontusmelke committed:
Fixes from code review

- Disable intersection scan for now
- Some stuff forgot to remove
- Some clean up

pontusmelke committed:
Update test mocks

pontusmelke committed:
Handle closing properly

pontusmelke committed:
Fix checkstyle

pontusmelke committed:
Disable union scans

We are not doing the correct thing when we only remove some of the labels
in the same transaction as we are doing the scan, say for example we
remove `:A` but do a union scan for `[:A, :B, :C]` in that case the result
depends on the state of the nodes on disk - if they have labels A or not.
